ml-proto/host/encode.ml

```diff
@@ -0,0 +1,401 @@
+      | F64_convert_s_i32 e -> unary e 0xae
+      | F64_convert_u_i32 e -> unary e 0xaf
+      | F64_convert_s_i64 e -> unary e 0xb0
+      | F64_convert_u_i64 e -> unary e 0xb1
+      | F64_promote_f32 e -> unary e 0xb2
+      | F64_reinterpret_i64 e -> unary e 0xb3
+      | I32_reinterpret_f32 e -> unary e 0xb4
+      | I64_reinterpret_f64 e -> unary e 0xb5
+
+    and unary e o = expr e; op o
+    and binary e1 e2 o = expr e1; expr e2; op o
+    and nary es o = list expr es; op o; arity es
+    and nary1 eo o = opt expr eo; op o; arity1 eo
+
+    (* Sections *)
+
+    let section id f x needed =
+      if needed then begin
+        let g = gap () in
+        let p = pos s in
+        string id;
+        f x;
+        patch_gap g (pos s - p)
+      end
+
+    (* Type section *)
+    let type_section ts =
+      section "type" (vec func_type) ts (ts <> [])
+
+    (* Import section *)
+    let import imp =
+      let {itype; module_name; func_name} = imp.it in
+      var itype; string module_name; string func_name
+
+    let import_section imps =
+      section "import" (vec import) imps (imps <> [])
+
+    (* Function section *)
+    let func f = var f.it.ftype
+
+    let func_section fs =
+      section "function" (vec func) fs (fs <> [])
+
+    (* Table section *)
+    let table_section tab =
+      section "table" (vec var) tab (tab <> [])
+
+    (* Memory section *)
+    let memory mem =
+      let {min; max; _} = mem.it in
+      vu64 min; vu64 max; bool true (*TODO: pending change*)
+
+    let memory_section memo =
+      section "memory" (opt memory) memo (memo <> None)
+
+    (* Export section *)
+    let export exp =
+      let {Kernel.name; kind} = exp.it in
+      (match kind with
+      | `Func x -> var x
+      | `Memory -> () (*TODO: pending resolution*)
+      ); string name
+
+    let export_section exps =
+      (*TODO: pending resolution*)
+      let exps = List.filter (fun exp -> exp.it.kind <> `Memory) exps in
+      section "export" (vec export) exps (exps <> [])
+
+    (* Start section *)
+    let start_section xo =
+      section "start" (opt var) xo (xo <> None)
+
+    (* Code section *)
+    let compress locals =
+      let combine t = function
+        | (t', n) :: ts when t = t' -> (t, n + 1) :: ts
+        | ts -> (t, 1) :: ts
+      in List.fold_right combine locals []
+
+    let local (t, n) = vu n; value_type t
+
+    let code f =
+      let {locals; body; _} = f.it in
+      vec local (compress locals);
+      let g = gap () in
+      let p = pos s in
+      list expr body;
+      patch_gap g (pos s - p)
+
+    let code_section fs =
+      section "code" (vec code) fs (fs <> [])
+
+    (* Data section *)
+    let segment seg =
+      let {Memory.addr; data} = seg.it in
+      vu64 addr; string data
+
+    let data_section segs =
+      section "data" (opt (vec segment))
+        segs (segs <> None && segs <> Some [])
+
+    (* End section *)
+    let end_section () =
+      section "end" ignore () true
```

Comment:
Currently there is no end section in BinaryEncoding.md.  If we want an explicit terminator, perhaps instead we can propose a `section_count : varuint32` that goes right after the version u32; but a dummy section seems like an abuse of what a section is.